File: experiments/java-sdk/src/main/java/com/datadog/llmobs/experiments/Evaluator.java
package com.datadog.llmobs.experiments;

/**
 * A user-supplied function that scores a single row's output.
 *
 * <p>The return type {@code V} drives the metric type sent to Datadog:
 * <ul>
 *   <li>{@link Boolean} -&gt; boolean metric
 *   <li>{@link Number}  -&gt; score metric (numeric)
 *   <li>anything else   -&gt; categorical metric (value stringified)
 * </ul>
 *
 * <p>Inputs are passed as {@code Object} — cast inside the lambda to match the types your
 * {@link Task} produces. An exception thrown here is recorded on the metric as an error and
 * does not abort the experiment.
 */
@FunctionalInterface
public interface Evaluator<V> {

    V evaluate(Object input, Object output, Object expectedOutput) throws Exception;
}

File: experiments/node-sdk/src/Evaluator.ts
/**
 * User-provided evaluator: `(input, output, expectedOutput) -> value`.
 *
 * May be sync or async. The return type drives the metric type (see MetricBuilder):
 *   - boolean -> boolean metric
 *   - number  -> score metric
 *   - anything else -> categorical metric (stringified)
 *
 * A thrown error / rejected promise is captured per-evaluation as an `error` on
 * the metric; the experiment continues.
 */
export type Evaluator<V = unknown> = (
  input: unknown,
  output: unknown,
  expectedOutput: unknown,
) => V | Promise<V>;
